src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java

```diff
@@ -126,62 +127,74 @@ public static long getKey0(int refIdx, int alignmentStart0) {
 		if(isInitialized)
 			close();
 		isInitialized = true;
+		reachedEnd = false;
 
 		final Configuration conf = ctx.getConfiguration();
 
 		final FileVirtualSplit split = (FileVirtualSplit)spl;
 		final Path             file  = split.getPath();
 		final FileSystem       fs    = file.getFileSystem(conf);
 
-		this.stringency = SAMHeaderReader.getValidationStringency(conf);
+		ValidationStringency stringency = SAMHeaderReader.getValidationStringency(conf);
 
-		final FSDataInputStream in = fs.open(file);
+		java.nio.file.Path index = SamFiles.findIndex(NIOFileUtil.asPath(fs.makeQualified(file).toUri()));
+		Path fileIndex = index == null ? null : new Path(index.toUri());
+		SeekableStream inStream = fileIndex == null ? null : WrapSeekable.openPath(fs, fileIndex);
+		in = WrapSeekable.openPath(fs, file);
+		SamReader samReader = createSamReader(in, inStream, stringency);
+		final SAMFileHeader header = samReader.getFileHeader();
 
-		final SAMFileHeader header = SAMHeaderReader.readSAMHeaderFrom(in, conf);
-		codec = new BAMRecordCodec(header);
-
-		in.seek(0);
-		bci =
-			new BlockCompressedInputStream(
-				new WrapSeekable<FSDataInputStream>(
-					in, fs.getFileStatus(file).getLen(), file));
-
-		virtualStart = split.getStartVirtualOffset();
+		long virtualStart = split.getStartVirtualOffset();
 
 		fileStart  = virtualStart >>> 16;
 		virtualEnd = split.getEndVirtualOffset();
 
-		bci.seek(virtualStart);
-		codec.setInputStream(bci);
-		
+		SamReader.PrimitiveSamReader primitiveSamReader =
+				((SamReader.PrimitiveSamReaderToSamReaderAdapter) samReader).underlyingReader();
+		BAMFileReader bamFileReader = (BAMFileReader) primitiveSamReader;
+
 		if(BAMInputFormat.DEBUG_BAM_SPLITTER) {
 			final long recordStart = virtualStart & 0xffff;
                 	System.err.println("XXX inizialized BAMRecordReader byte offset: " +
 				fileStart + " record offset: " + recordStart);
 		}
 
-		keepReadPairsTogether = SortOrder.queryname.equals(header.getSortOrder()) &&
-			conf.getBoolean(BAMInputFormat.KEEP_PAIRED_READS_TOGETHER_PROPERTY, false);
-		readPair = false;
-		lastOfPair = false;
-		intervals = BAMInputFormat.getIntervals(conf);
+		List<Interval> intervals = BAMInputFormat.getIntervals(conf);
```

Comment:
I'm wondering if we should throw if we see a call that has the KEEP_PAIRED_READS_TOGETHER_PROPERTY set, since otherwise we're silently no longer honoring it.